editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -452,58 +507,87 @@ export function fixJSXElementChildUIDs(
           whenTrue: updatedWhenTrue,
           whenFalse: updatedWhenFalse,
         })
+      } else {
+        return updateUID(jsxConditionalExpressionUIDOptic, oldElement.uid, fixUIDsState, newElement)
       }
-      break
     }
     case 'ATTRIBUTE_VALUE':
     case 'ATTRIBUTE_NESTED_ARRAY':
     case 'ATTRIBUTE_NESTED_OBJECT':
     case 'ATTRIBUTE_FUNCTION_CALL':
     case 'ATTRIBUTE_OTHER_JAVASCRIPT': {
-      if (oldElement.type === newElement.type) {
+      if (oldElement == null || oldElement.type === newElement.type) {
         return fixExpressionUIDs(oldElement, newElement, fixUIDsState)
+      } else {
+        return updateUID(jsExpressionUIDOptic, oldElement.uid, fixUIDsState, newElement)
       }
-      break
     }
     default:
       assertNever(newElement)
   }
-
-  return newElement
 }
 
 export function fixJSXElementUIDs(
-  oldElement: JSXElement,
+  oldElement: JSXElementChild | null | undefined,
   newElement: JSXElement,
   fixUIDsState: FixUIDsState,
 ): JSXElement {
-  const fixedProps = fixJSXAttributesUIDs(oldElement.props, newElement.props, fixUIDsState)
-  const fixedChildren = fixJSXElementChildArray(
-    oldElement.children,
-    newElement.children,
-    fixUIDsState,
-  )
-  return updateUID(jsxElementUIDOptic, oldElement.uid, fixUIDsState, {
-    ...newElement,
-    props: fixedProps,
-    children: fixedChildren,
-  })
+  if (oldElement == null) {
+    // No previous entry, so just return the new value.
+    return newElement
+  } else {
+    // Update the UID upfront.
+    const elementWithUpdatedUID = updateUID(
+      jsxElementUIDOptic,
+      oldElement.uid,
+      fixUIDsState,
+      newElement,
+    )
+
+    // Set the `data-uid` attribute.
+    const attributesWithUpdatedUID: JSXAttributes = setJSXAttributesAttribute(
+      elementWithUpdatedUID.props,
+      'data-uid',
+      jsExpressionValue(elementWithUpdatedUID.uid, emptyComments),
+    )
+
+    // If this is a `JSXElement`, then work through the common fields.
+    let fixedProps: JSXAttributes = attributesWithUpdatedUID
+    let fixedChildren: Array<JSXElementChild> = elementWithUpdatedUID.children
+    if (isJSXElement(oldElement)) {
+      fixedProps = fixJSXAttributesUIDs(oldElement.props, attributesWithUpdatedUID, fixUIDsState)
+      fixedChildren = fixJSXElementChildArray(
+        oldElement.children,
+        elementWithUpdatedUID.children,
+        fixUIDsState,
+      )
+    }
+
+    return {
+      ...elementWithUpdatedUID,
+      props: fixedProps,
+      children: fixedChildren,
+    }
+  }
 }
 
 export function fixExpressionUIDs(
-  oldExpression: JSExpression,
+  oldExpression: JSExpression | null | undefined,
   newExpression: JSExpression,
   fixUIDsState: FixUIDsState,
 ): JSExpression {
   switch (newExpression.type) {
     case 'ATTRIBUTE_VALUE': {
-      if (oldExpression.type === newExpression.type) {
+      if (oldExpression == null) {
+        return newExpression
```

Comment:
minor, but I think this could be extracted to the top of the function (since it's repeated in all the case branches)

Reply:
Good shout, fixed in 44f39ba48.